Add tests for ASGI specification compliance

LucidDan opened the pull request:
WIP: test suite for ASGI specification compliance.

It seems like a great idea for a ASGI test library to have tests itself
to check it is compliant with the spec, and if there are new versions
that come out, will make it easier to update to match the spec.

There are a number of tests that currently xfail in this suite - 
one or two of those tests might be things that I'm wrong in my
interpretation of the spec, but I'm mostly pretty confident in them.

I've intentionally made these xfails rather than just fixing the code in
the same PR, so the idea would be to fix the issues one by one and
remove the xfail, or in cases where the issue is a WONTFIX, change
the test and/or indicate as such in the xfail reason.

There is still a need for a couple more tests in the websocket
section, and probably could do with some more type-oriented
negative tests (e.g. testing what happens if you send a message
with the wrong type in the event values, like with 'status').

Of particular note, the scope['path'] test (test_http:test_http_path_is_not_escaped)
I'm not entirely sure that test should exist, it might not be valid.
More explanation in the comments. This one is one of the main things
that makes me want to keep this PR draft a bit longer, both for comment
and while I decide what to do with that test.

LucidDan committed:
Add tests for ASGI specification compliance

There are a number of tests that currently xfail in this suite.
The intention is to fix the issues one by one, or in cases where
the issue is a WONTFIX, change the test and/or indicate as such
in the xfail reason.

There is still a need for a couple more tests in the websocket
section, and probably could do with some more type-oriented
negative tests (e.g. testing what happens if you send a message
with the wrong type in the event values, like with 'status')

LucidDan committed:
Allow http.response.body event with no 'body' key

Fixes test_http:test_response_body_can_be_missing()

LucidDan committed:
Allow http.response.start event with no 'headers' key

Fixes test_http:test_response_headers_can_be_missing()

LucidDan committed:
Ensure websocket events include 'asgi' in scope

Fixes test_websocket :
 test_asgi_version_is_present_in_websocket_scope() and
 test_websocket_spec_version_is_missing_or_correct()

LucidDan committed:
Disable lifespan protocol if apps do not support it

Fixes test_lifespan:test_lifespan_not_supported_is_allowed()

This fix was somewhat challenging to achieve without breaking
other things or backward compatibility issues.

In the end, I think the most elegant solution is to adopt
more Pythonic exceptions in the TestClient, as demonstrated in
the commit. However, since this is a less trivial change, I
have only used TestClientError in the lifespan related code,
 leaving all other code using raw Exceptions still.

I think TestClientError (or possibly even multiple exception classes
for different purposes) should actually be adopted throughout
all the code, albeit with the caveat that it might cause some issues
if anyone is doing really rigid exception assertions in their test
code. For most cases (including my own code that is already using
async_asgi_testclient) I believe the average
 'with pytest.raises(Exception)' assertions should still work,
 since TestClientError is a sub-class of Exception.

A side-benefit of the TestClientError is it would allow access to
the original Message structure, instead of having to decode a string
repr of the event message.

LucidDan committed:
Remove test for custom scope

Removed test for contamination between calls with a custom scope.
Probably deserves documentation because it could be a gotcha for people,
but it doesn't deserve to be in this test suite.

LucidDan commented:
The CI black is pretty old, and I can't get it to run on my computer at all (symbol not found in flat namespace '__PyUnicode_DecodeUnicodeEscape')...so I can't get this to pass. :-(

I broke the commits out so it is easy to read through and see the fixes for each xfail test, but they don't necessarily need to be split out like that; might be better to just squash them when merging.

LucidDan committed:
Fix formatting

Had a few issues with formatting and the black version, this commit cleans that up.

LucidDan commented:
This PR now includes fixes for a few of the most prominent issues I wrote tests for (read: the ones personally causing me immediate problems right now :-D).

The main one that deserves the most attention is the one that originally sent me down this rabbit hole - the lifespan protocol handling. When finding a solution for that, I considered a few approaches, but I ended up using a custom exception - I would actually recommend using TestClientError everywhere instead of the current use of raw Exceptions which is generally considered bad practice. The commit has some detailed notes on the thinking behind it.

masipcat commented:
Hi @LucidDan, I've done a quick review and it looks really good! I'll do a proper review when I come back from the holiday in about two weeks. Let me know when this is no longer a WIP and it's ready for review.

LucidDan commented:
Great!
Yeah it's no longer a WIP, I think there is a little more that could be done (I don't have any test for websocket send/receive) but otherwise it is in good shape.

masipcat commented on async_asgi_testclient/testing.py:
```diff
@@ -78,33 +79,46 @@ def __init__(
+        except TestClientError:
+            # Pass these through directly, so that test clients can assert on them
+            raise
+        except:  # noqa
+            # Any other exception is (almost) definitely passed through from the app under test
+            # So it means the lifespan protocol is not supported.
+            self._lifespan_task = None
```
Maybe we could log the exception just in case. I prefer not to silence exceptions entirely without any feedback. Also, I guess it's important we notify the user that lifespan protocol is not supported

LucidDan replied:
Yeah that seems reasonable, I'll add it. This is the behaviour I'd expect many real ASGI server would have (to warn the user), so we should behave the same way.

masipcat commented on async_asgi_testclient/tests/asgi_spec/conftest.py:
```diff
@@ -0,0 +1,175 @@
+        # Receive http.requests until http.disconnect or more_body = False
+        while True:
+            msg.append(await receive())
+            if msg[-1]["type"] == "http.disconnect" or not msg[-1].get(
+                "more_body", False
+            ):
```
Should we check that if have not received a `http.disconnect`, it is a `http.request`?

Like:

```suggestion
            if msg[-1]["type"] == "http.disconnect" or (msg[-1] == "http.request" and not msg[-1].get(
                "more_body", False
            )):
```

So we can catch unexpected events sent by asgi-testclient.

LucidDan replied:
Yes this makes sense. That whole function could probably be rewritten to be more readable too, now that I look at it after a break. ;-)

masipcat replied:
great!

masipcat commented on async_asgi_testclient/tests/asgi_spec/test_http.py:
```diff
@@ -0,0 +1,391 @@
+        assert resp.status_code == 200
+
+
+@pytest.mark.xfail(
+    AssertionError, reason="TestClient does not uppercase scope['method']"
+)
```
Good catch! I think we should fix it. Did you try to fix it and see if the integration test with some frameworks still pass?

LucidDan replied:
No, I didn't touch this one - I thought this was up to you as maintainer to decide when is the right time if it will changed, because it will potentially break existing usage, so you might want to schedule it for an appropriate minor version with a breaking change notice or whatever makes sense for the package.

masipcat replied:
I'm planning to release your changes in v2.0.0 and document these breaking changes in the CHANGELOG. 

I guess all the major framworks already uppercase the HTTP method, and that's why this is working even though it's not correct with the specification.

masipcat commented on async_asgi_testclient/tests/asgi_spec/test_http.py:
```diff
@@ -0,0 +1,391 @@
+
+
+# Ok, this one I'm not sure about...
+# The grey area here is TestClient doesn't currently decode an encoded URL; but should it?
```
I think the TestClient should encode the `path` internally, the same way we are doing for the `query_string`:

https://github.com/vinissimus/async-asgi-testclient/blob/56779ef69ff2eebd88178980d6f8ba50b91b1529/async_asgi_testclient/utils.py#L49-L53

If you want to change it and add/modify one of  the integration tests to ensure the frameworks still work I think it's enough :)

LucidDan committed:
Merge branch 'vinissimus:master' into master

LucidDan committed:
Notify the user if lifespan protocol raised an exception

Re: https://github.com/vinissimus/async-asgi-testclient/pull/50/files#r785889284

The asserts where we access the log message in the test is a bit rough, it could probably be refactored to be a little cleaner.

LucidDan committed:
Fix method is uppercased in HTTP

LucidDan commented:
Been a long time, but I'm taking another look at this...still a couple of things to resolve but have addressed a few of the questions from the (Jan 2022!!) review.
